In [26]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
import nltk
import re
import nltk
from nltk.corpus import stopwords


In [5]:
#Please comment out the dataset you don't want
df = pd.read_excel('booking_sentiment.xlsx',sheet_name='summary')
#df = pd.read_excel('TwitterSentiment.xlsx',sheet_name='summary')


In [6]:
df
X = df['message']
y = df['label']

In [7]:
#Clean the number
X = X.apply(lambda x: re.sub('\d','num',x.lower()))
#Clean the punctuation
X = X.apply(lambda x: re.sub('\-|\,|\.|\!',' ',x.lower()))
X

0      location perfect  very nice and attentive staf...
1                           good location  helpful staff
2      great value for money nice swimming pool nice ...
3      i am pretty tall and big man and here is one o...
4      the location was excellent   the restraint was...
                             ...                        
134    so bad  we have joined for four nights but onl...
135                                  gym was a bit small
136    the only thing i can say is not so good is the...
137    pool towels disgusting have thrown better in t...
138    very poor repesentation of by the hotel staff ...
Name: message, Length: 139, dtype: object

In [8]:
#Download stop word from nltk (So we don't need to define it ourselve)
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /Users/euer/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [9]:
print(stopwords.words('english'))

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

In [10]:
stopwords = stopwords.words('english')

In [23]:
document = list(X)
type(document)

list

In [27]:
cv_doc = CountVectorizer(document)

TypeError: __init__() takes 1 positional argument but 2 were given

In [13]:
 cv_vector = cv_doc.fit_transform(document)

NameError: name 'cv_doc' is not defined

In [ ]:
list(cv_doc.get_feature_names())

['absolutely',
 'ac',
 'acceptable',
 'access',
 'accomodating',
 'across',
 'action',
 'add',
 'additional',
 'adelphi',
 'adequate',
 'adjust',
 'adjusted',
 'adults',
 'aerier',
 'afternoons',
 'aging',
 'air',
 'aircon',
 'allowed',
 'allows',
 'almost',
 'also',
 'always',
 'amateur',
 'amenities',
 'amount',
 'apparently',
 'approach',
 'approx',
 'area',
 'areas',
 'armost',
 'around',
 'arrived',
 'asia',
 'asian',
 'attendant',
 'attentive',
 'available',
 'average',
 'away',
 'back',
 'bad',
 'baht',
 'balcony',
 'bangkok',
 'bar',
 'bars',
 'basic',
 'bath',
 'bathroom',
 'beach',
 'bed',
 'bedroom',
 'beside',
 'best',
 'better',
 'beyond',
 'big',
 'bin',
 'bit',
 'bizarre',
 'blurry',
 'body',
 'book',
 'booking',
 'bored',
 'bottom',
 'breakfast',
 'breakfasts',
 'bring',
 'brother',
 'buakhao',
 'buffet',
 'building',
 'bukhao',
 'bukhoew',
 'bus',
 'busy',
 'calm',
 'camp',
 'cannot',
 'card',
 'care',
 'center',
 'central',
 'certain',
 'certainly',
 'chair',
 'change

In [ ]:
print(cv_vector)

  (0, 312)	1
  (0, 411)	1
  (0, 366)	1
  (0, 38)	1
  (0, 528)	1
  (0, 103)	1
  (0, 243)	1
  (0, 53)	1
  (0, 18)	1
  (1, 312)	1
  (1, 528)	1
  (1, 243)	1
  (1, 262)	1
  (2, 312)	1
  (2, 366)	2
  (2, 245)	1
  (2, 609)	1
  (2, 347)	1
  (2, 553)	1
  (2, 425)	1
  (2, 144)	1
  (2, 471)	1
  (2, 357)	1
  (2, 308)	1
  (2, 419)	1
  :	:
  (138, 97)	1
  (138, 537)	1
  (138, 265)	2
  (138, 214)	1
  (138, 616)	1
  (138, 626)	1
  (138, 543)	1
  (138, 367)	2
  (138, 577)	1
  (138, 450)	1
  (138, 426)	1
  (138, 131)	1
  (138, 33)	1
  (138, 597)	1
  (138, 219)	1
  (138, 52)	1
  (138, 455)	1
  (138, 297)	1
  (138, 474)	1
  (138, 510)	1
  (138, 207)	1
  (138, 487)	1
  (138, 257)	1
  (138, 138)	1
  (138, 292)	1


<img src = 'sum_axis_0.jpg'>

In [ ]:
sum_word = {cv_doc.get_feature_names()[i]:v for i,v in enumerate(cv_vector.toarray().sum(axis=0))}

In [ ]:
#Show the most mentioned words from imput (excluded stopwords)
sorted(sum_word.items(),key= lambda item:item[1],reverse=True)

[('good', 57),
 ('staff', 46),
 ('breakfast', 41),
 ('room', 36),
 ('hotel', 35),
 ('location', 35),
 ('clean', 28),
 ('pool', 24),
 ('nice', 21),
 ('friendly', 18),
 ('great', 18),
 ('num', 16),
 ('helpful', 15),
 ('rooms', 12),
 ('could', 11),
 ('night', 11),
 ('also', 10),
 ('like', 10),
 ('service', 10),
 ('shower', 10),
 ('swimming', 10),
 ('numnum', 9),
 ('value', 9),
 ('well', 9),
 ('bars', 8),
 ('bed', 8),
 ('money', 8),
 ('noisy', 8),
 ('pattaya', 8),
 ('bit', 7),
 ('close', 7),
 ('due', 7),
 ('everything', 7),
 ('outside', 7),
 ('star', 7),
 ('street', 7),
 ('walk', 7),
 ('walking', 7),
 ('water', 7),
 ('air', 6),
 ('area', 6),
 ('best', 6),
 ('check', 6),
 ('cleaning', 6),
 ('excellent', 6),
 ('far', 6),
 ('gym', 6),
 ('loud', 6),
 ('minutes', 6),
 ('nights', 6),
 ('poor', 6),
 ('quiet', 6),
 ('restaurants', 6),
 ('would', 6),
 ('better', 5),
 ('big', 5),
 ('cold', 5),
 ('comfortable', 5),
 ('enough', 5),
 ('little', 5),
 ('many', 5),
 ('much', 5),
 ('one', 5),
 ('pretty', 5

In [ ]:
#This will be input(X) accordingly
cv_vector

<139x639 sparse matrix of type '<class 'numpy.int64'>'
	with 1512 stored elements in Compressed Sparse Row format>

# Modelling

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.metrics import confusion_matrix,accuracy_score
from sklearn.model_selection import train_test_split

In [ ]:
X  = cv_vector

In [ ]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2,random_state=10)

In [ ]:
#Model to be compare

dt = DecisionTreeClassifier()
svc = SVC()
nb = GaussianNB()
mlp = MLPClassifier()
ada = AdaBoostClassifier()
ensemble_clf = [dt,svc,nb,mlp,ada]

In [ ]:
#For model nb, there is some adjustment for X needed
for model in [dt,svc,nb,mlp,ada]:
    if model != nb:
        clf = model.fit(X_train,y_train)
        globals()[f'y_pred_{model}'] = clf.predict(X_test)
        print(f'y_pred_{model} accuracy = ',accuracy_score(globals()[f'y_pred_{model}'],y_test))
    else:
        clf = model.fit(X_train.toarray(),y_train)
        globals()[f'y_pred_{model}'] = clf.predict(X_test.toarray())
        print(f'y_pred_{model} accuracy = ',accuracy_score(globals()[f'y_pred_{model}'],y_test))
    

y_pred_DecisionTreeClassifier() accuracy =  0.8214285714285714
y_pred_SVC() accuracy =  0.9285714285714286
y_pred_GaussianNB() accuracy =  0.8928571428571429
y_pred_MLPClassifier() accuracy =  0.8928571428571429
y_pred_AdaBoostClassifier() accuracy =  0.8571428571428571


# Twitter Performance

<img src='Twitter_performance.png' width=480>

# Booking Performance

<img src='Booking_performance.png' width = 480>

# Tuning hyperparameter with GridCV

In [ ]:
params1={"max_depth": range(5,30,5), "min_samples_leaf": range(1,30,2)}
params2={"kernel":['sigmoid','linear','rbf']}
params3={}
params4={'solver': ['lbfgs'], 'max_iter': [1000], 'alpha': 10.0 ** -np.arange(1, 10), 'hidden_layer_sizes':np.arange(10, 15),'learning_rate' :['constant','adaptive']}
params5={'n_estimators':[50,100,150,200],'learning_rate':np.arange(0.1,1,0.1)}
parameters_list=[params1, params2, params3, params4,params5]

In [ ]:
model_log = ['_dt','_svc','_nb','_mlp','_ada']

In [ ]:
#For model nb, there is some adjustment for X needed
for i in range(len(ensemble_clf)):
    if i != 2:
        clf = GridSearchCV(estimator = ensemble_clf[i],param_grid=parameters_list[i],cv=3).fit(X_train,y_train)
        globals()[f'y_pred_{ensemble_clf[i]}'] = clf.predict(X_test)
        print(f'y_pred_{ensemble_clf[i]} best params: ',clf.best_params_)
        print(f'y_pred_{ensemble_clf[i]} accuracy = ',accuracy_score(globals()[f'y_pred_{ensemble_clf[i]}'],y_test))
    else:
        clf = GridSearchCV(estimator = ensemble_clf[i],param_grid=parameters_list[i],cv=3).fit(X_train.toarray(),y_train)
        globals()[f'y_pred_{ensemble_clf[i]}'] = clf.predict(X_test.toarray())
        print(f'y_pred_{ensemble_clf[i]} best params: ',clf.best_params_)
        print(f'y_pred_{ensemble_clf[i]} accuracy = ',accuracy_score(globals()[f'y_pred_{ensemble_clf[i]}'],y_test))
    

y_pred_DecisionTreeClassifier() best params:  {'max_depth': 20, 'min_samples_leaf': 1}
y_pred_DecisionTreeClassifier() accuracy =  0.8214285714285714
y_pred_SVC() best params:  {'kernel': 'sigmoid'}
y_pred_SVC() accuracy =  0.9285714285714286
y_pred_GaussianNB() best params:  {}
y_pred_GaussianNB() accuracy =  0.8928571428571429
y_pred_MLPClassifier() best params:  {'alpha': 0.001, 'hidden_layer_sizes': 14, 'learning_rate': 'constant', 'max_iter': 1000, 'solver': 'lbfgs'}
y_pred_MLPClassifier() accuracy =  0.8928571428571429
y_pred_AdaBoostClassifier() best params:  {'learning_rate': 0.6, 'n_estimators': 150}
y_pred_AdaBoostClassifier() accuracy =  0.8571428571428571


# Conclusion
1. Performance
   - Twitter dataset : Naive Bayes (GB) performed the best with 85.7% accuracy
   - Booking dataset : SVC performed the best with 92.8% accuracy
   - The performance is quite good interms of accuracy but that because of the test set unique words are included in the trainning set already </br>
   ,so there is no words that the training set haven't seen which make the performance so good.
   - In my opinion, this is very surprising that the performance is this high, especially for booking dataset, however, the set unique words in the data </br>
   set seems to be very narrow </br></br>
2. Improvement
   - As the sentiment analysis need lots more datapoints for the sake of having bigger bag of words which provide more realiable result.
   - More adjustment for data preprocessing such as ignore the outlier, drop the mispell might help enhacing the performance.